launcher/src/main/java/org/apache/spark/launcher/LauncherServer.java

```diff
@@ -331,23 +358,27 @@ protected void handle(Message msg) throws IOException {
           timeout.cancel();
         }
         close();
+        if (handle != null) {
```

Comment:
This is the fix for the new race described in the summary (the code is moved here from below).

This changes behavior slightly: the handle now waits for the child process / thread to finish before disposing itself, whereas before that would happen as soon as the connection between the child process / thread was closed.